source/opt/decoration_manager.h

```diff
@@ -35,10 +36,16 @@ class DecorationManager {
   }
   DecorationManager() = delete;
 
-  // Removes all decorations from |id|, which should not be a group ID.
-  void RemoveDecorationsFrom(uint32_t id);
+  // Removes all decorations from |id| for which |pred| returns true.
+  // |id| should not be a group ID.
+  void RemoveDecorationsFrom(uint32_t id,
```

Comment:
Can you explain why you are adding `pred`?  It will make the implementation much more difficult, so we should have a reason.

Reply:
The reason for adding pred was explained in #898, more specifically in @dneto0’s comment https://github.com/KhronosGroup/SPIRV-Tools/issues/898#issuecomment-338413513. Currently the linker has some code (which is wrong) that loops over all decorations to remove only LinkageAttributes Import ones, or FuncParamAttr ones. The idea is to have the DecorationManager do it, in a generic way, instead.